python/chemiscope/jupyter.py

```diff
@@ -33,16 +33,33 @@ class ChemiscopeWidgetBase(ipywidgets.DOMWidget, ipywidgets.ValueWidget):
     # timeout for warning messages
     warning_timeout = Int(10000).tag(sync=True)
 
-    def __init__(self, data, has_metadata, warning_timeout=10000):
+    def __init__(
+        self, data, has_metadata, warning_timeout=10000, cache_structures=True
+    ):
         super().__init__()
-        self.value = json.dumps(data)
         self.has_metadata = has_metadata
         if "settings" in data:
             self.settings = data["settings"]
         self._settings_sync = True
         # timeout for warning messages (ms). 0 to make persistent, -1 to disable
         self.warning_timeout = warning_timeout
 
+        # hold structures on the python side to save js memory
+        self._structures_cache = {}
+        if cache_structures and "data" not in data["structures"][0]:
+            structure_refs = []
+            for i, structure in enumerate(data["structures"]):
+                if "data" in structure:
+                    raise ValueError("Mixed explicit and external structures in frames")
```

Comment:
if the 5th structure has "data" in it, structures 0-4 will already be cached before the error is raised. properly, we should validate the entire array first before caching anything... but idk if it might introduce some overhead